include/fixed_containers/fixed_robinhood_hashtable.hpp

```diff
@@ -0,0 +1,574 @@
+        }
+        bucket_at(table_loc) = bucket;
+    }
+
+    constexpr void do_erase(SizeType table_loc)
+    {
+        SizeType value_index = bucket_at(table_loc).value_index_;
+ 
+        // shift down until either empty or an element with correct spot is found
+        SizeType next_loc = next_bucket_index(table_loc);
+        while(bucket_at(next_loc).dist_and_fingerprint_ >= Bucket::DIST_INC * 2)
+        {
+            bucket_at(table_loc) = bucket_at(next_loc).minus_dist();
+            table_loc = std::exchange(next_loc, next_bucket_index(next_loc));
+        }
+        bucket_at(table_loc) = {};
+
+        // if the item being removed isn't the last item in the value array, swap that into this slot
+        if(value_index != value_array_size() - 1)
+        {
+            // move the last element over the deleted value
+            SizeType last_idx = value_array_size() - 1;
+            value_at(value_index) = std::move(value_at(last_idx));
+
+            // we need to find the bucket that currently points at this value and fix the pointer
+            // we do this separately from normal search logic because we know the `value_index_` we're looking for instead of being able to compare stuff _in_ `value_array_`
+            uint64_t h = hash(get_key(value_at(value_index)));
+            table_loc = bucket_index_from_hash(h);
+            while(last_idx != bucket_at(table_loc).value_index_)
+            {
+                table_loc = next_bucket_index(table_loc);
+            }
+            bucket_at(table_loc).value_index_ = value_index;
+        }
+        IMPLEMENTATION_DETAIL_DO_NOT_USE_value_array_.pop_back();
+    }
+
+    template <typename Key>
+    constexpr bool do_erase_key(const Key& key)
+    {
+        if(value_array_size() == 0)
+        {
+            return false;
+        }
+
+        // find the first bucket that might contain this key
+        auto [dist_and_fingerprint, table_loc] = next_while_less(key);
+
+        // it is possible that this bucket contains a hash collision instead of the right value. If so, we have to iterate through all the values that might match
+        while(dist_and_fingerprint == bucket_at(table_loc).dist_and_fingerprint_
+            && !key_equal(get_key(value_at(bucket_at(table_loc).value_index_)), key))
+        {
+            dist_and_fingerprint = Bucket::increment_dist(dist_and_fingerprint);
+            table_loc = next_bucket_index(table_loc);
+        }
+
+        // we exited the while loop by finding a bucket that no longer matches, so we didn't find the key
+        if (dist_and_fingerprint != bucket_at(table_loc).dist_and_fingerprint_)
+        {
+            return false;
+        }
+
+        // we found it, do the actual erase
+        do_erase(table_loc);
+        return true;
+    }
+
+    template <typename Key, typename... Args>
+    constexpr SizeType do_emplace(uint32_t dist_and_fingerprint, SizeType table_loc, Key&& key, Args&&... args)
```

Comment:
We can have a minimal set of functions in the hashtable and let `FixedUnorderedMap`/`Set` have the richer API.
Also, names should match the red_black_tree implementation where appropriate. (Can use the names here and replace the existing ones)